### PR TITLE
refact(): storage now provides async operations.

### PR DESCRIPTION
**Pull request by antmat**
@3Hren PTAL

### CODE REVIEW

File: src/runtime/runtime.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ main(int argc, char* argv[]) {
         ("configuration,c", po::value<std::string>(), "location of the configuration file")
         ("logging,l", po::value<std::string>()->default_value("core"), "logging backend")
 #if !defined(__APPLE__)
-        ("daemonize,d", "daemonize on start")
+    unable to decode message arguments        ("daemonize,d", "daemonize on start")
```
Comment (3Hren) [inline src/runtime/runtime.cpp]: WTF?

File: src/storage/files.cpp [hunk 2]
```diff
@@ -48,16 +48,16 @@ files_t::~files_t() {
     // Empty.
 }
 
-std::string
-files_t::read(const std::string& collection, const std::string& key) {
+void
+files_t::read(callback<std::string> cb, const std::string& collection, const std::string& key) {
```
Comment (3Hren) [inline src/storage/files.cpp]: We should prefer asynchronous computations here. Via thread-pool for example.

File: include/cocaine/api/storage.hpp [hunk 3]
```diff
@@ -44,82 +44,98 @@ struct storage_t {
...
-        std::unique_lock<std::mutex> lock(m);
-        f([&](std::future<R> future) {
-            result = std::move(future);
-            cv.notify_one();
-        }, args...);
-        cv.wait(lock);
-        return result.get();
+    template <class T>
+    static
+    void assing_future_result(std::promise<T>& promise, std::future<T>& future) {
```
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: Why lvalue reference instead of value semantics here?

File: include/cocaine/api/storage.hpp [hunk 4]
```diff
@@ -44,82 +44,98 @@ struct storage_t {
...
+        try {
+            promise.set_value(future.get());
+        } catch (...) {
+            promise.set_exception(std::current_exception());
+        }
+    }
+
+    static
+    void assing_future_result(std::promise<void>& promise, std::future<void>& future) {
```
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: And here.

File: include/cocaine/api/storage.hpp [hunk 5]
```diff
@@ -129,55 +145,51 @@ storage_t::get(callback<T> cb, const std::string& collection, const std::string&
...
-    read(std::move(inner_cb), collection, key);
+    read(collection, key, std::move(inner_cb));
+}
+
+template<class T>
+std::future<T>
+storage_t::get(const std::string& collection, const std::string& key) {
+    auto promise = std::make_shared<std::promise<T>>();
+    get<T>(collection, key, [=](std::future<T> future){
+        assing_future_result(*promise, future);
```
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: `std::move(future)` here.

File: include/cocaine/api/storage.hpp [hunk 6]
```diff
@@ -129,55 +145,51 @@ storage_t::get(callback<T> cb, const std::string& collection, const std::string&
...
-template<class T>
-void
-storage_t::put_sync(const std::string& collection, const std::string& key, const T& object, const std::vector<std::string>& tags) {
-    namespace ph = std::placeholders;
-    wrap_sync<void>(std::bind(&storage_t::put<T>, this, ph::_1, ph::_2, ph::_3, ph::_4, ph::_5), collection, key, object, tags);
+std::future<void>
+storage_t::put(const std::string& collection, const std::string& key, const T& object, const std::vector<std::string>& tags) {
+    std::promise<void> promise;
+    put(collection, key, object, tags, [=](std::future<T> future) mutable {
+        assing_future_result(promise, future);
```
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: AFAIK `promise` isn't copyable, is it?

File: src/storage/files.cpp [hunk 7]
```diff
@@ -41,126 +42,124 @@ using blackhole::attribute_list;
...
+    io_loop(),
+    io_work(io_loop),
+    thread([&](){
+        io_loop.run();
+    })
 { }
 
 files_t::~files_t() {
-    // Empty.
+    io_work = boost::none;
```
Comment (3Hren) [inline src/storage/files.cpp]: Just `io_work.reset()`.

File: src/storage/files.cpp [hunk 8]
```diff
@@ -41,126 +42,124 @@ using blackhole::attribute_list;
...
+    io_loop.post([=](){
+        const fs::path file_path(m_parent_path / collection / key);
 
-    if(!fs::exists(file_path)) {
-        return cb(make_exceptional_future<std::string>(std::make_error_code(std::errc::no_such_file_or_directory),
-            file_path.string()
-        ));
-    }
+        if(!fs::exists(file_path)) {
+            auto ec = std::make_error_code(std::errc::no_such_file_or_directory);
```
Comment (3Hren) [inline src/storage/files.cpp]: Look, domain and asynchronous logic intersection here. Isn't it better to decompose file sync reading/writing into separate functions, that throw exceptions (like it was before) and then make them private and provide async overloads, which does only:

```c++
loop.post([&]() {
    try { 
        cb(read(...)); 
    catch (...) {
        cb(make_exceptional_future<std::string>(std::current_exception());
    }
})
```

?

Comment (3Hren) [inline src/storage/files.cpp]: Moreover then we're drop indents, that hardens code review.

File: include/cocaine/utility/future.hpp [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+#pragma once
+
+#include <future>
+#include <string>
+#include <system_error>
+
+namespace cocaine {
+
+template<class T>
+std::future<T> make_ready_future(T value) {
```
Comment (3Hren) [inline include/cocaine/utility/future.hpp]: Hmm! Fix here before merge!

**Comment by 3Hren** [comment]
The build has failed.

**Comment by antmat** [comment]
Damn boost::optional

**Comment by antmat** [comment]
Looks like it's ok, except plugins now.

**Comment by antmat** [comment]
oops, one more fix

**Pull request merged**